src/txn/txn.c

```diff
@@ -149,28 +149,27 @@ __wt_txn_user_active(WT_SESSION_IMPL *session)
     WT_CONNECTION_IMPL *conn;
     WT_SESSION_IMPL *session_in_list;
     WT_TXN_GLOBAL *txn_global;
-    WT_TXN_SHARED *txn_shared;
     uint32_t i, session_cnt;
     bool txn_active;
 
     conn = S2C(session);
     txn_active = false;
     txn_global = &conn->txn_global;
 
+    WT_STAT_CONN_INCR(session, txn_walk_sessions);
+
     /* We're going to scan the table: wait for the lock. */
     __wt_writelock(session, &txn_global->rwlock);
 
     WT_ORDERED_READ(session_cnt, conn->session_cnt);
-    WT_STAT_CONN_INCR(session, txn_walk_sessions);
-    for (i = 0, session_in_list = conn->sessions, txn_shared = txn_global->txn_shared_list;
-         i < session_cnt; i++, session_in_list++, txn_shared++) {
+    for (i = 0, session_in_list = conn->sessions; i < session_cnt; i++, session_in_list++) {
 
         /* Skip inactive or internal sessions. */
         if (!session_in_list->active || F_ISSET(session_in_list, WT_SESSION_INTERNAL))
             continue;
 
         /* Check if a user session has a running transaction. */
-        if (txn_shared->id != WT_TXN_NONE || txn_shared->pinned_id != WT_TXN_NONE) {
+        if (F_ISSET(session_in_list->txn, WT_TXN_RUNNING)) {
```

Comment:
What makes me nervous here is accessing a transaction attribute that isn't in the shared transaction attributes list. Is that OK because we are holding the transaction global lock?

Reply:
Actually, this is just reading through the list of active sessions, we don't need any kind of lock for that at all (and I've just pushed that cleanup). Of course, ignore that change if this approach just doesn't work.

Obviously, we can race with a thread starting a transaction, but that's true by definition for this function, I don't see how we can get into trouble there.

Is there any way we can be here where a session doesn't have a transaction structure? I don't think so, but I'm not 100% sure.

Reply:
@keithbostic There are some problems in accessing the session->txn structure directly. When this function is originally implemented as part of WT-7909 it used the session->txn structure directly and it leads to problems and later it is rewritten with txn_shared values in WT-8203.

Reply:
@kommiharibabu, I can dig if you don't know off-hand, but if you do know, what are those problems?

Reply:
@agorrod, @kommiharibabu filled me in in a separate channel. The failure is BF-22434 and it was fixed in WT-8023. In summary, RTS can race with session deactivation which is freeing the `WT_SESSION.WT_TXN` structure that's read in this check.

It seems to me there are two alternatives: fix the locking so we can't race with session deactivation, or add a way to tell from the underlying shared txn structure that the transaction is in a prepared state. The former seems easier and cleaner to me, so that's what I did.

@kommiharibabu, if you think updating the shared information or another path is a better choice, I'm happy to defer to you on this question.

Reply:
There's no way for a session to not have a transaction structure; it's part of the session allocation, and it never changes at runtime either.

However, the pending diffs for reading from checkpoints do change it temporarily. I don't think the resulting race is likely to cause any problems. But that change will itself likely break this check -- with those diffs you could prepare a transaction, then go off and read from a checkpoint, and the prepared transaction won't be seen in the session while this is happening. That can be fixed in the checkpoint-reading branch (though not for free, requires checks when accessing session->txn) but it might be better to use txn_shared.

Reply:
(I wrote something here and deleted it, please disregard)